### PR TITLE
Choose the precision for the histogram ticks dynamically

### PR DESCRIPTION
**Pull request by akopich**
Instead of using the hard-coded number of decimals we
choose the minimal number of decimals s.t. all the ticks
have different values. We have to limit the maximal number of
decimals due to space limitations.

Further, the tick is now rendered at the position driven by the
rounded value.

After rounding some ticks may fall off to the right or to the left.
Therefore, the maximal number of tics has been increased by 1.

Test: Manual
Bug: http://b/224967376

### CODE REVIEW

**Comment by akopich** [comment]
http://screenshot/8bbLP5cdgH3bBZi

File: src/OrbitQt/HistogramWidget.cpp
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:159]: making a double_variable++ is a bit strange to me. 

Can't you directly return here?

File: src/OrbitQt/HistogramWidget.cpp
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:203]: Nit: Maybe adding a comment before, "skipping ticks outside the visible axis range" or something like that.

File: src/OrbitQt/HistogramWidget.cpp
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:297]: Add '{' and '}' now that the statement occupies two lines.

File: src/OrbitQt/HistogramWidget.cpp
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:306]: Why you are erasing this? 
And why you are not erasing the closing one?

**Review by karupayun — APPROVED** [review]
Hi Valeriy, I left a few comments. Most of them nits.

File: src/OrbitQt/HistogramWidget.cpp
Comment (akopich) [reply src/OrbitQt/HistogramWidget.cpp:159]: I can. Good point.

File: src/OrbitQt/HistogramWidget.cpp
Comment (akopich) [reply src/OrbitQt/HistogramWidget.cpp:306]: I have only moved it all the way up.

File: src/OrbitQt/HistogramWidget.cpp
Comment (karupayun) [reply src/OrbitQt/HistogramWidget.cpp:98]: I see. I would have them together generating directly the pairs (there are not too many labels so this can be a problem in any way), but I don't have a strong opinion about that, it's just I don't like to have two vectors that had a hidden relationship. I would rather use some kind of map (labels, values)

**Review by karupayun — APPROVED** [review]
LGTM!

**Comment by dpallotti** [comment]
> After rounding some ticks may fall off to the right or to the left.
Therefore, the maximal number of tics has been increased by 1.

> http://screenshot/8bbLP5cdgH3bBZi

Is it common that the rightmost value on the horizontal axis and the top value on the vertical axis are not visible? I don't really like the look.
My opinion would be that the last tick determines the height and width of the histogram, a bit past the maximum values. I guess this is not what you had in mind and I don't know how big of a code change that would be though.

File: src/OrbitQt/HistogramWidget.cpp
Comment (dpallotti) [inline src/OrbitQt/HistogramWidget.cpp:96]: This should be in an anonymous namespace.

**Comment by akopich** [comment]
@dpallotti Carsten's UX proposal does allow for the missing ticks on the edges of axes. After discussion we've agreed that the last tick doesn't bring value in the presence of the hovering labels. 

Adding the last tick "a bit past the maximum values" could be an option, yet it would make the histogram bars smaller and I wouldn't like to waste the space. Also, it feels like a significant code change.

**Comment by dpallotti** [comment]
> WDYT? http://screenshot/8fGGTxianXWDHxH http://screenshot/8aJHFr7u3o7bUnZ

Being completely honest, the second screenshot here makes me prefer the initial solution...

**Comment by karupayun** [comment]
> > After rounding some ticks may fall off to the right or to the left.
> > Therefore, the maximal number of tics has been increased by 1.
> 
> > http://screenshot/8bbLP5cdgH3bBZi
> 
> Is it common that the rightmost value on the horizontal axis and the top value on the vertical axis are not visible? I don't really like the look. My opinion would be that the last tick determines the height and width of the histogram, a bit past the maximum values. I guess this is not what you had in mind and I don't know how big of a code change that would be though.

+1

**Comment by karupayun** [comment]
I strongly think that both horizontal ticks and vertical ticks should follow an arithmetic sequence.


> @dpallotti fair enough. What if we round to smallest tick up and largest tick down? (I'd also keep the code that discards the ticks falling out, as it's not clear to me, why this can't happen now). WDYT? http://screenshot/8fGGTxianXWDHxH http://screenshot/8aJHFr7u3o7bUnZ
> 
> @karupayun I've also changed the code, now the string-labels and rounded values are instantiated together.

Sorry, I didn't realize that distance between labels was not constant. Maybe I should create a bug, but I'm strongly in disagree with having labels which are not an arithmetic sequence. I've just ping Carsten (which is near me) and he has the same opinion.

**Review by karupayun — CHANGES_REQUESTED** [review]
As I explained above, I changed my mind and I don't agree to have labels which don't follow an arithmetic sequence.

**Comment by akopich** [comment]
@karupayun right now we have equidistant ticks with equal labels (see the bug). 
Making the ticks equidistant, keeping their number low enough and choosing the sufficient number of decimal points (at the same time) seems to require disproportionate amount of effort.

**Comment by akopich** [comment]
@karupayun there's an option. But it will have a drawback of sometimes having just two ticks. Do we care?

**Comment by karupayun** [comment]
> @karupayun there's an option. But it will have a drawback of sometimes having just two ticks. Do we care?

No, actually if I force you to always have 3 visible ticks I think the problem became impossible.

**Comment by akopich** [comment]
Carsten has implemented the idea in JS, convincing me to give it a try. 

http://screenshot/C8x3ut3Jb2SH7qH

**Comment by karupayun** [comment]
> 
> http://screenshot/C8x3ut3Jb2SH7qH

I love it :), will review it as soon as possible but you have my visual approval :)

File: src/OrbitQt/HistogramWidget.cpp
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:122]: Nit: A comment to explain this method? (for example to explain that we are not always going to get the desired number of ticks)

**Review by karupayun — APPROVED** [review]
LGTM! Thanks for addressing all my comments :)

File: src/OrbitQt/HistogramWidget.cpp
Comment (akopich) [reply src/OrbitQt/HistogramWidget.cpp:52]: So the variable is initialised;

File: src/OrbitQt/HistogramWidget.cpp
Comment (akopich) [reply src/OrbitQt/HistogramWidget.cpp:129]: IMO one tick is just a bad as zero ticks. Neither gives any sense of scale. At least, we've got hover labels.

Comment (akopich) [reply src/OrbitQt/HistogramWidget.cpp:129]: But I don't see why choosing 0 over 1 tick is impossible in this implementation either. Probably, it is.

**Pull request merged**
